cmsl1t/io/eventreader.py

```diff
@@ -0,0 +1,135 @@
+                      - event.myAlias
+        to
+        {'myAlias': (<tree path>, <branchname>.<leafname>)}
+
+    '''
+    # TODO: move this replacement into cmsl1t.config
+    aliasMap = {}
+    trees = ntuple_map['content']
+    for treeName, content in trees.items():
+        for branchName, branchContent in content['branches'].items():
+            for alias in branchContent['aliases']:
+                tidyAlias = alias.replace('event.', '')
+                aliasMap[tidyAlias] = (treeName, branchName)
+    return aliasMap
+
+
+class EventReader(object):
+
+    def __init__(self, input_files, ntuple_map, nevents=-1):
+        '''
+            Reads ntuple_info as defined by bin/create-map-file
+        '''
+        self._treeNames = ntuple_map['content'].keys()
+        self._aliasMap = _create_alias_map(ntuple_map)
+        self.input_files = _get_input_files(input_files)
+        self.nevents = nevents
+        self._trees = {}
+        for alias in self._aliasMap:
+            if 'vertex' in alias.lower():
+                print(alias)
+
+        self._load_trees()
+
+    def _load_trees(self):
+        for treeName in self._treeNames:
+            try:
+                self._trees[treeName] = TreeChain(
+                    treeName,
+                    self.input_files,
+                    cache=True,
+                    events=self.nevents,
+                )
+            except RuntimeError:
+                logger.warn(
+                    "Cannot find tree: {0} in input file".format(treeName))
+                continue
+
+    def contains(self, name):
+        return name in self._aliasMap.keys()
+
+    def __iter__(self):
+        # event loop
+        for trees in six.moves.zip(*self._trees.itervalues()):
+            yield Event(self._trees, self._aliasMap)
+
+
+class Event(object):
+
+    energySumTypes = {
+        'Ett': sumTypes.kTotalEt,
+        'EttHF': sumTypes.kTotalEtHF,
+        'Htt': sumTypes.kTotalHt,
+        'HttHF': sumTypes.kTotalHtHF,
+        'Met': sumTypes.kMissingEt,
+        'MetHF': sumTypes.kMissingEtHF,
+        'Mht': sumTypes.kMissingHt,
+        'Mex': sumTypes.kTotalEtx,
+        'Mey': sumTypes.kTotalEty,
+    }
+
+    def __init__(self, trees, mapping):
+        self._map = mapping
+        self._trees = trees
+        self._cache = {}
+
+    def __getattr__(self, name):
+        if name in object.__getattribute__(self, '_cache'):
+            return object.__getattribute__(self, '_cache')[name]
+
+        if name not in object.__getattribute__(self, '_map'):
+            return object.__getattribute__(self, name)
+        treeName, treeAttr = object.__getattribute__(self, '_map')[name]
+        tree = object.__getattribute__(self, '_trees')[treeName]
+        obj = tree
+        for attr in treeAttr.split('.'):
+            obj = getattr(obj, attr)
+        object.__getattribute__(self, '_cache')[name] = obj
+        return obj
+
+    def __getitem__(self, name):
```

Comment:
I'd be tempted to only implement `__getattr__`.  Having them both implemented seems a bit unnecessary and makes it a bit less clear, plus if we ever find a reason why they should behave differently we'd have a tough time unpicking the code that uses this.

Reply:
Because you can use dataframe-like syntax:
```python
# in analzer, filling histograms
h.Fill(event['MET'])
```
can work with both single entries (event-for-loop) and data columns (e.g. rootpy histograms).

Furthermore, `__getitem__` is just a pointer to `__getattr__`.